server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java

```diff
@@ -201,6 +201,22 @@ NumericType numericType() {
          */
         public abstract long roundUpToMillis(long value);
 
+        public final void validateTimestamp(long value, DocumentParserContext context) {
+            if (context.indexSettings().getTimeSeriesStartTime() == null) {
+                assert context.indexSettings().getTimeSeriesEndTime() == null;
+                return;
+            }
+            long startTime = convert(context.indexSettings().getTimeSeriesStartTime());
+            if (value < startTime) {
+                throw new IllegalArgumentException("time series index @timestamp value [" + value + "] must be larger than " + startTime);
+            }
+
+            long endTime = convert(context.indexSettings().getTimeSeriesStartTime());
+            if (value >= endTime) {
+                throw new IllegalArgumentException("time series index @timestamp value [" + value + "] must be smaller than " + endTime);
+            }
+        }
```

Comment:
This method makes it fairly simple to handle millis and nanos using the "native" resolution.